fuel-vm/src/interpreter/memory.rs

```diff
@@ -287,6 +327,75 @@ where
     }
 }
 
+pub(crate) fn push_selected_registers(
+    memory: &mut [u8; MEM_SIZE],
+    mut sp: RegMut<SP>,
+    hp: Reg<HP>,
+    pc: RegMut<PC>,
+    program_regs: &ProgramRegisters,
+    segment: ProgramRegistersSegment,
+    bitmask: Imm24,
+) -> Result<(), RuntimeError> {
+    let bitmask = bitmask.to_u32();
+
+    // First compute the new stack pointer, as that's the only error condition
+    let count = bitmask.count_ones();
+    let stack_range = MemoryRange::new(*sp, (count as u64) * 8)?;
+    let new_sp = stack_range.words().end;
+    if new_sp > *hp {
+        return Err(PanicReason::MemoryOverflow.into())
+    }
+
+    // Write the registers to the stack
+    let mut it = memory[stack_range.usizes()].chunks_exact_mut(8);
+    for (i, reg) in program_regs.segment(segment).iter().enumerate() {
+        if (bitmask & (1 << i)) != 0 {
+            it.next().unwrap().copy_from_slice(&reg.to_be_bytes());
+        }
+    }
+
+    // Apply changes to system registers
+    *sp = new_sp;
+    inc_pc(pc)
+}
+
+pub(crate) fn pop_selected_registers(
+    memory: &[u8; MEM_SIZE],
+    mut sp: RegMut<SP>,
+    ssp: Reg<SSP>,
+    pc: RegMut<PC>,
+    program_regs: &mut ProgramRegisters,
+    segment: ProgramRegistersSegment,
+    bitmask: Imm24,
+) -> Result<(), RuntimeError> {
+    let bitmask = bitmask.to_u32();
+
+    // First compute the new stack pointer, as that's the only error condition
+    let count = bitmask.count_ones();
+    let size_in_stack = (count as u64) * 8;
+    let new_sp = sp
+        .checked_sub(size_in_stack)
+        .ok_or(PanicReason::MemoryOverflow)?;
+    if new_sp < *ssp {
+        return Err(PanicReason::MemoryOverflow.into())
```

Comment:
Is this an overflow or MemoryOwnership error?

Reply:
Overflow, since `MemoryOwnership` is would mean violation of the ownership rules. Stack and heap overlap isn't exactly that. Ideally we'd have another error here, called `OutOfMemory` or something like that. There are several other places where clarifying the panic reason would be nice, so I think leaving that to a follow-up PR would be the correct call in case we want to have that.